source/elements/oneDAL/source/algorithms/clustering/kmeans.rst

```diff
@@ -1,153 +1,494 @@
-         .set_cluster_count(10);
+   .. member:: kmeans::model model = kmeans::model()
+
+      The trained K-means model.
+
+      Getter
+         | ``const model& get_model() const``
+
+
+   .. member:: table labels = table()
+
+      :math:`n \times 1` table with assignments of cluster indices to samples in
+      the input data.
+
+      Getter
+         | ``const table& get_labels() const``
+
+
+   .. member:: std::int64_t iteration_count = 0
+
+      Actual number of iterations performed by the algorithm.
+
+      Invariants
+         | :expr:`iteration_count >= 0`
+
+
+   .. member:: double objective_function_value = 0.0
+
+      The value of the objective function :math:`\Phi_X(C)`, where :math:`C` is
+      :expr:`model.centroids` (see :expr:`kmeans::model::centroids`).
+
+      Invariants
+         | :expr:`objective_function_value >= 0.0`
 
-      onedal::train(kmeans_desc, x_train, x_train_centroids);
-   }
 
+Operation semantics
+~~~~~~~~~~~~~~~~~~~
+.. namespace:: onedal
+.. function:: template <typename Descriptor> \
+              kmeans::train_result train(const Descriptor& desc, \
+                                         const kmeans::train_input& input)
 
-Inference
----------
+   :tparam Descriptor: K-Means algorithm descriptor :expr:`kmeans::desc`.
+
+   Preconditions
+      | :expr:`input.data.is_empty == false`
+      | :expr:`input.initial_centroids.is_empty == false`
+      | :expr:`input.initial_centroids.row_count == desc.cluster_count`
+      | :expr:`input.initial_centroids.column_count == input.data.column_count`
+
+   Postconditions
+      | :expr:`result.labels.is_empty == false`
+      | :expr:`result.labels.row_count == input.data.row_count`
+      | :expr:`result.model.centroids.is_empty == false`
+      | :expr:`result.model.clusters.row_count == desc.cluster_count`
+      | :expr:`result.model.clusters.column_count == input.data.column_count`
+      | :expr:`result.iteration_count <= desc.max_iteration_count`
+
+
+Inference :expr:`onedal::infer(...)`
+------------------------------------
 
 Input
 ~~~~~
-.. code-block:: c++
-
-   namespace onedal::kmeans {
+::
 
    class infer_input {
    public:
       infer_input();
       infer_input(const model& m);
       infer_input(const model& m, const table& data);
 
+      const model& get_model() const;
+      const table& get_data() const;
+
       infer_input& set_model(const model&);
       infer_input& set_data(const table&);
    };
 
-   } // namespace onedal::kmeans
+.. namespace:: onedal::kmeans
+.. class:: infer_input
+
+   .. function:: infer_input()
+
+      Creates input for the inference operation with the default attribute
+      values.
+
+
+   .. function:: infer_input(const kmeans::model& model)
+
+      Creates input for the inference operation with the given :expr:`model`, the
+      other attributes get default values.
+
+
+   .. function:: infer_input(const kmeans::model& model, const table& data)
```

Comment:
generally, later need double-check style, spaces, etc

Reply:
It's HTML generator problem, RST is ok